lib/Provider/Command/MessageSend.php

```diff
@@ -0,0 +1,155 @@
+	 * @param array $options array of options reserved for future use
+	 *
+	 * @return LocalMessage
+	 *
+	 * @throws SendException on failure, check message for reason
+	 *
+	 */
+	public function perform(string $userId, string $serviceId, IMessage $message, array $options = []): LocalMessage {
+		// validate that at least one To address is present
+		if (count($message->getTo()) === 0) {
+			throw new SendException('Invalid Message Parameter: MUST contain at least one TO address with a valid address');
+		}
+		// validate that all To, CC and BCC have email address
+		$entries = array_merge($message->getTo(), $message->getCc(), $message->getBcc());
+		array_walk($entries, function ($entry) {
+			if (empty($entry->getAddress())) {
+				throw new SendException('Invalid Message Parameter: All TO, CC and BCC addresses MUST contain at least an email address');
+			}
+		});
+		// validate that all attachments have a name, type, and contents
+		$entries = $message->getAttachments();
+		array_walk($entries, function ($entry) {
+			if (empty($entry->getName()) || empty($entry->getType()) || empty($entry->getContents())) {
+				throw new SendException('Invalid Attachment Parameter: MUST contain values for Name, Type and Contents');
+			}
+		});
+		// retrieve user mail account details
+		try {
+			$account = $this->accountService->find($userId, (int)$serviceId);
+		} catch (ClientException $e) {
+			throw new SendException('Error: occurred while retrieving mail account details', 0, $e);
+		}
+		// convert mail provider message to mail app message
+		$localMessage = new LocalMessage();
+		$localMessage->setType($localMessage::TYPE_OUTGOING);
+		$localMessage->setAccountId($account->getId());
+		$localMessage->setSubject((string)$message->getSubject());
+		$localMessage->setBody((string)$message->getBody());
+		// disabled due to issues caused by opening these messages in gui
+		//$localMessage->setEditorBody($message->getBody());
+		$localMessage->setHtml(true);
+		$localMessage->setSendAt($this->time->getTime());
+		// convert mail provider addresses to recipient addresses
+		$to = $this->convertAddressArray($message->getTo());
+		$cc = $this->convertAddressArray($message->getCc());
+		$bcc = $this->convertAddressArray($message->getBcc());
+		// save attachments
+		$attachments = [];
+		try {
+			foreach ($message->getAttachments() as $entry) {
+				$attachments[] = $this->attachmentService->addFileFromString(
+					$userId,
+					(string)$entry->getName(),
+					(string)$entry->getType(),
+					(string)$entry->getContents()
+				)->jsonSerialize();
+			}
+		} catch (UploadException $e) {
+			$this->purgeSavedAttachments($attachments);
+			throw new SendException('Error: occurred while saving mail message attachment', 0, $e);
+		}
+		// save message
+		try {
+			$localMessage = $this->outboxService->saveMessage(
+				$account,
+				$localMessage,
+				$to,
+				$cc,
+				$bcc,
+				$attachments
+			);
+		} catch (\Throwable $e) {
```

Comment:
I could not find any documented exception to be thrown by `saveMessage`. Remove the try-catch.

Reply:
I left this in one in on purpose, the saveMessage function actually spawns a Imap connection, so I though it would be a good idea to remove the local attachments if the function fails. 

Removed